### PR TITLE
Allow interaction with hyperlinks in mouse mode

### PR DESCRIPTION
**Pull request by Don-Vito**
## PR Checklist
* [x] Closes https://github.com/microsoft/terminal/issues/9117
* [x] CLA signed. 
* [x] Tests added/passed
* [ ] Documentation updated.
* [ ] Schema updated.
* [x] I've discussed this with core contributors already. 

## Detailed Description of the Pull Request / Additional comments
In mouse mode:
* Underline hyperlinks
* Activate hyperlink on ctrl+click rather than sending input to VT

### CODE REVIEW

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -1347,7 +1346,16 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
                     !(_terminal->GetHyperlinkAtPosition(terminalPosition).empty()))
                 {
                     _HyperlinkHandler(_terminal->GetHyperlinkAtPosition(terminalPosition));
+                    args.Handled(true);
+                }
+
+                if (args.Handled())
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.cpp:1352]: unrelated: mind moving the comment from 1344 down to 1358

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -1427,10 +1435,8 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
             {
                 _TrySendMouseEvent(point);
                 args.Handled(true);
-                return;
             }
-
-            if (_focused && point.Properties().IsLeftButtonPressed())
+            else if (_focused && point.Properties().IsLeftButtonPressed())
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.cpp:1439]: this all seems unrelated. But if we're doing this, I think you can remove 1437 too? We set args.Handled to true on line 1519.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 3]
```diff
@@ -1347,7 +1346,16 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
                     !(_terminal->GetHyperlinkAtPosition(terminalPosition).empty()))
                 {
                     _HyperlinkHandler(_terminal->GetHyperlinkAtPosition(terminalPosition));
+                    args.Handled(true);
+                }
+
+                if (args.Handled())
+                {
+                    // We already handled the event by either sending it to the VT,
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.cpp:1354]:
```suggestion
                    // GH#9117: We already handled the event by either sending it to the VT,
```
nit: link the issue for easier tracking

**Review by carlos-zamora — APPROVED** [review]

**Comment by zadjii-msft** [comment]
wait is this going to immediately conflict with #9403?

**Comment by Don-Vito** [comment]
> 
> 
> wait is this going to immediately conflict with #9403?

Code-wise yes :) Logically not :)

**Comment by Don-Vito** [comment]
@zadjii-msft - but before reviewing it further - probably it worth considering the discussion in the ticket (#9117). Probably we want to switch the links to ctrl+shift+click to be consistent between mouse mode and non-mouse mode without doing the hacks I did (of both passing the click and handling the link).

**Comment by zadjii-msft** [comment]
Hmm. I'm not sure how I feel about that. I like the idea of consistency, but I dislike the idea of adding shift to the default modifiers you need to press.

If I'm looking to other terminal emulators for precedent, it seems like hyperlinks in `gnome-terminal` are actually a _higher_ priority than mouse mode. So you can click on links with just ctrl+click even in mouse mode. Maybe we should follow that precedent? I'll ask the team

**Comment by zadjii-msft** [comment]
Okay yea team consensus was that the `gnome-terminal` behavior was sensible and we should just do that. So no need for shift while clicking links with mouse mode enabled - the hyperlink takes priority over the mouse mode click

**Comment by Don-Vito** [comment]
@zadjii-msft, @carlos-zamora -please re-review :blush:

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 4]
```diff
@@ -1351,6 +1344,11 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
                 {
                     _HyperlinkHandler(_terminal->GetHyperlinkAtPosition(terminalPosition));
                 }
+                else if (_CanSendVTMouseInput())
+                {
+                    // GH#9396: hyper-link gets higher priority than mouse mode
+                    _TrySendMouseEvent(point);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:1350]: I'm a bit worried about the flow of code through these conditionals. This will change the block selection state, track click counts, etc. before dispatching a mouse event.

We might be overdue for refactoring the click dispatcher. . . but this seems a bit too entangled for my taste?

Comment (Don-Vito) [reply src/cascadia/TerminalControl/TermControl.cpp]: @DHowett you are right as usual, currently the hyperlink handling was already entangled into the selection processing code. This forced me to push the VT processing there as well (as hyperlink should precede). 

I issued another commit, trying to decouple everything into:
1. try to handle hyperlink
2. else try to handle VT mouse
3. else try to update selection (left-click)
4. else try to handle paste (right-click)

**Review by zadjii-msft — APPROVED** [review]

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 5]
```diff
@@ -1308,20 +1308,27 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
-            if (_CanSendVTMouseInput())
+            auto lock = _terminal->LockForWriting();
+            const auto cursorPosition = point.Position();
+            const auto terminalPosition = _GetTerminalPosition(cursorPosition);
+
+            // GH#9396: we prioritize hyper-link over VT mouse events
+            if (point.Properties().IsLeftButtonPressed() && ctrlEnabled && !_terminal->GetHyperlinkAtPosition(terminalPosition).empty())
+            {
+                // Handle hyper-link only on the first click to prevent multiple activations
+                auto clickCount = _NumberOfClicks(cursorPosition, point.Timestamp());
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.cpp:1319]:
```suggestion
                const auto clickCount = _NumberOfClicks(cursorPosition, point.Timestamp());
```

Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControl.cpp:1319]: Also, could we just move this outside of the if-statement over by line 1313? That way we don't do it again on line 1335?

Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControl.cpp:1319]: My only concern with that would be that we're updating the click count whenever we try to send VT Mouse Input, but I think that's ok. Mind playing with that a little bit to make sure though?

**Review by carlos-zamora — APPROVED** [review]
This is such a nit, so I'll approve anyways. But I won't put automerge on it until I get a response. Thanks!

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 5 above)
Comment (Don-Vito) [reply src/cascadia/TerminalControl/TermControl.cpp:1319]: @carlos-zamora - yes, I didn't put the clicks count earlier to avoid counting the number of clicks in VT mouse mode. I am not sure how important it is, but Dustin was somewhat worried with that in my original solution.

Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControl.cpp:1319]: Fair enough. Can I at least get a `const` on the two `clickCount`s then? :P

Comment (Don-Vito) [reply src/cascadia/TerminalControl/TermControl.cpp:1319]: > 
> 
> Fair enough. Can I at least get a `const` on the two `clickCount`s then? :P

Lol! Absolutely! Doing it :)

Comment (Don-Vito) [reply src/cascadia/TerminalControl/TermControl.cpp]: @carlos-zamora - applied the consts. Thanks for the feedback!
